tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs

```diff
@@ -189,6 +189,25 @@ type AsyncModule() =
             |> ignore
             if c.Value = 2 then Assert.Fail("both error and cancel continuations were called")
 
+    [<Fact>]
+    member _.``Async.RunImmediate should execute on the same thread`` () =
+        let t =
+            async {
+                let a = Thread.CurrentThread.ManagedThreadId
+                let b =
+                    async {
+                        return Thread.CurrentThread.ManagedThreadId
+                    } |> Async.RunImmediate
+                let c = Thread.CurrentThread.ManagedThreadId
+                return $"Before: {a}, in async: {b}, after async: {c}"
+            } |> Async.RunImmediate
+
+        let d = Thread.CurrentThread.ManagedThreadId
+        let actual = $"{t}, after task: {d}"
+
+        if not (actual = $"Before: {d}, in async: {d}, after async: {d}, after task: {d}") then
+            failwith actual
```

Comment:
Took me a while to understand what's going on here, I'd just give vars clear names ("threadIdBefore", "threadIdAfter"...) and assert them all against `Thread.CurrentThread.ManagedThreadId`.